grakn-graph/src/main/java/ai/grakn/graph/internal/ConceptLog.java

```diff
@@ -40,9 +43,20 @@
  *
  */
 public class ConceptLog {
-    private Set<ConceptImpl> modifiedConcepts;
+
+    //Caches any concept which has been touched before
+    private final Map<ConceptId, ConceptImpl> conceptCache = new HashMap<>();
```

Comment:
This might be a good place to use that guava cache: https://github.com/google/guava/wiki/CachesExplained